### PR TITLE
txn: add a new field `txn_source` in write and lock

### PR DESCRIPTION
**Pull request by xiongjiwei**
### What is changed and how it works?

Issue Number: ref https://github.com/tikv/tikv/issues/13779

What's Changed:

```commit-message
This PR is the starting preparation of BDR, and only adds the ability to parse and serialize `txn_source`
```

### Check List 

Tests 

- Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by MyonKeminta — COMMENTED** [review]
It seems (de)serialization of Lock/Write records with non-zero `txn_source` is not covered by tests. You can consider add some records with non-zero `txn_source` in `test_lock` and `test_write`.

File: components/txn_types/src/write.rs [hunk 1]
```diff
@@ -409,6 +437,9 @@ impl WriteRef<'_> {
         if !self.last_change_ts.is_zero() {
             size += 1 + size_of::<u64>() + MAX_VAR_U64_LEN;
         }
+        if self.txn_source != 0 {
```
Comment (sticnarf) [inline components/txn_types/src/write.rs:440]: It looks better if you write code consistently. Use `!= 0` or `> 0` in both lock and write..

**Review by sticnarf — COMMENTED** [review]
Can you write some tests for 1pc and 2pc and check whether the txn_source is written as expected?

File: components/txn_types/src/write.rs [hunk 2]
```diff
@@ -354,6 +372,11 @@ impl WriteRef<'_> {
                     last_change_ts = number::decode_u64(&mut b)?.into();
                     versions_to_last_change = number::decode_var_u64(&mut b)?;
                 }
+                TXN_SOURCE_PREFIX => {
+                    txn_source = b
+                        .read_u8()
+                        .map_err(|_| Error::from(ErrorInner::BadFormatWrite))?
```
[the remark this answers is not in the document]
Comment (xiongjiwei) [reply components/txn_types/src/write.rs:378]: it has to, the error returned should be an inner error, I follow the other code in this function

**Review by sticnarf — COMMENTED** [review]
Mostly lgtm. Please fix the compile errors.

**Review by sticnarf — APPROVED** [review]
LGTM

**Review by MyonKeminta — APPROVED** [review]

**Pull request merged**